File: frontend/src/api/contracts.ts
import client from './client';
import type { Contract, ContractListResponse } from '../types';

export async function listContracts(params?: {
  search?: string;
  status?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: string;
}): Promise<ContractListResponse> {
  const res = await client.get('/contracts', { params });
  return res.data;
}

export async function createContract(data: {
  title: string;
  contract_no: string;
  party_a: string;
  party_b: string;
  amount?: number;
  sign_date?: string;
  expiry_date?: string;
  content?: string;
}): Promise<Contract> {
  const res = await client.post('/contracts', data);
  return res.data;
}

export async function getContract(id: number): Promise<Contract> {
  const res = await client.get(`/contracts/${id}`);
  return res.data;
}

export async function updateContract(
  id: number,
  data: {
    title?: string;
    contract_no?: string;
    party_a?: string;
    party_b?: string;
    amount?: number;
    sign_date?: string | null;
    expiry_date?: string | null;
    content?: string;
  }
): Promise<Contract> {
  const res = await client.put(`/contracts/${id}`, data);
  return res.data;
}

export async function deleteContract(id: number): Promise<void> {
  await client.delete(`/contracts/${id}`);
}

export async function updateContractStatus(
  id: number,
  status: string
): Promise<Contract> {
  const res = await client.patch(`/contracts/${id}/status`, { status });
  return res.data;
}

File: backend/app/routers/contracts.py
"""Contract management API endpoints."""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models import User
from app.schemas.contract import (
    ContractCreate,
    ContractUpdate,
    ContractResponse,
    ContractListResponse,
    ContractStatusUpdate,
)
from app.services import contract_service

router = APIRouter(prefix="/api/contracts", tags=["contracts"])


@router.get("", response_model=ContractListResponse)
def list_contracts(
    search: str | None = Query(None, description="搜索合同编号/标题/甲乙方"),
    status: str | None = Query(None, description="按状态筛选"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = Query("id"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List contracts with search, filter, pagination."""
    return contract_service.list_contracts(
        db, current_user, search, status, page, page_size, sort_by, sort_order
    )


@router.post("", response_model=ContractResponse, status_code=status.HTTP_201_CREATED)
def create_contract(
    body: ContractCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new contract (default status: draft)."""
    try:
        contract = contract_service.create_contract(db, body, current_user)
        return contract_service._to_response(contract)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.get("/{contract_id}", response_model=ContractResponse)
def get_contract(
    contract_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get contract detail."""
    try:
        return contract_service.get_contract(db, contract_id, current_user)
    except ValueError as e:
        if "不存在" in str(e):
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))


@router.put("/{contract_id}", response_model=ContractResponse)
def update_contract(
    contract_id: int,
    body: ContractUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update contract information."""
    try:
        return contract_service.update_contract(db, contract_id, body, current_user)
    except ValueError as e:
        if "不存在" in str(e):
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        if "无权" in str(e):
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.delete("/{contract_id}")
def delete_contract(
    contract_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete a contract (draft only)."""
    try:
        contract_service.delete_contract(db, contract_id, current_user)
        return {"message": "合同已删除"}
    except ValueError as e:
        if "不存在" in str(e):
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        if "无权" in str(e):
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.patch("/{contract_id}/status", response_model=ContractResponse)
def update_contract_status(
    contract_id: int,
    body: ContractStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Transition contract status (draft→active→terminated/expired)."""
    try:
        return contract_service.update_contract_status(
            db, contract_id, body.status, current_user
        )
    except ValueError as e:
        if "不存在" in str(e):
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        if "无权" in str(e):
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
